server/routes.go

```diff
@@ -197,20 +202,32 @@ func GenerateHandler(c *gin.Context) {
 
 	checkpointLoaded := time.Now()
 
-	prompt := req.Prompt
-	if !req.Raw {
-		prompt, err = model.Prompt(req)
+	var prompt string
+	sendContext := false
+	switch {
+	case req.Raw:
+		prompt = req.Prompt
+	case areMessagesSet:
+		prompt, err = promptFromMessages(model, req.Messages)
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+			return
+		}
+	case req.Prompt != "":
+		prompt, err = promptFromRequestParams(c, model, req)
 		if err != nil {
 			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
 			return
 		}
+		sendContext = true
 	}
 
+	var generated strings.Builder
 	ch := make(chan any)
 	go func() {
 		defer close(ch)
 		// an empty request loads the model
-		if req.Prompt == "" && req.Template == "" && req.System == "" {
+		if req.Prompt == "" && req.Template == "" && req.System == "" && !areMessagesSet {
```

Comment:
While better if we don't have to, we could add some helper functions to the request struct. e.g. `req.Empty()`